# Data and AI Assignment 2
![UCLL logo](https://cdn.discordapp.com/attachments/1025724154852417549/1033683765530218506/unknown.png)

Group
- Astrid Piot
- Catarina Rilho
- Cristina Marques
- Greg François Reynders
- Robbe Van de Velde
- Seppe Haems

## Bussiness understanding
---


### 1) Determine the business objectives 

    A school wants to graduate as many high quality students as possible and have them occupy important positions in the workfield. 
    This means keeping the standard of graduates high, while also passing a lot of students. 
    Having talented students at their disposal makes this job easier. 
    Thus A school needs to keep its reputation high, in order to convince talented students to apply for their   
    school, international students and local ones. A delicate balance between passing a lot of students, and keeping the bar high emerges. 
   
### 2) Assess situation: What needs to be done to get ready?
      
    Risk: We are not allowed to make decision's based on race or gender
    Todo: gather student, and course data. 
    Cost: will be zero in terms of currency, we only need to request data. 
 
### 3) Determine data science goals
      
    Given data about students we want to predict if they are going to be successful or not.
    Analyzing which courses are failed/passed the most, in order to improve these courses.
    How do students interact with the virtual learning environments and how does it improve their chances to succeed in the courses.
      
### 4) Produce project plan
      
    This would take me about 1 hour, I will research the data and find patterns.


## Data understanding
---

### 1) Collect intial data

    Already provided


### 2) Examine the data

#### Dictionary

##### Assessments

*This file contains information about assessments in module-presentations. Usually, every presentation has a number of assessments followed by the final exam.*

|variable                       |class          |description                    |notes                        |
|:------------------------------|:--------------|:------------------------------|:----------------------------|
|code_module                    |character      |identification code of the module||
|code_presentation              |character      |identification code of the presentation||
|id_assessment                  |number         |identification number of the assessment||
|assessment_type                |character      |type of assessment. Three types of assessments exist: Tutor Marked Assessment (TMA), Computer Marked Assessment (CMA) and Final Exam (Exam)||
|date                           |number         |information about the final submission date of the assessment|We don't really know how to interpret it. Would need further clarification from the business.|
|weight                         |double     |weight of the assessment in %. Typically, Exams are treated separately and have the weight 100%; the sum of all other assessments is 100%||

##### Courses

*File contains the list of all available modules and their presentations. The columns are:*

|variable                       |class          |description                    |notes                        |
|:------------------------------|:--------------|:------------------------------|:----------------------------|
|code_module                    |character      |code name of the module, which serves as the identifier||
|code_presentation              |character      |code name of the presentation. It consists of the year and “B” for the presentation starting in February and “J” for the presentation starting in October.||
|module_presentation_length     |number         |length of the module-presentation in days||

*The structure of B and J presentations may differ and therefore it is good practice to analyse the B and J presentations separately. Nevertheless, for some presentations the corresponding previous B/J presentation do not exist and therefore the J presentation must be used to inform the B presentation or vice versa. In the dataset this is the case of CCC, EEE and GGG modules.*


##### Student Assessment

*This file contains the results of students’ assessments. If the student does not submit the assessment, no result is recorded. The final exam submissions is missing, if the result of the assessments is not stored in the system. This file contains the following columns:*

|variable                       |class          |description                    |notes                        |
|:------------------------------|:--------------|:------------------------------|:----------------------------|
|id_assessment                  |number         |the identification number of the assessment||
|id_student                     |number         |a unique identification number for the student||
|date_submitted                 |number         |the date of student submission||
|is_banked                      |boolean        |a status flag indicating that the assessment is banked|0  or 1, I also don't really know what this means|
|score                          |double         |the student’s score in this assessment|0 to 100|

##### Student Registration

*This file contains information about the time when the student registered for the module presentation. For students who unregistered the date of unregistration is also recorded. File contains five columns:*

|variable                       |class          |description                    |notes                        |
|:------------------------------|:--------------|:------------------------------|:----------------------------|
|code_module                    |character      |an identification code for a module||
|code_presentation              |character      |the identification code of the presentation||
|id_student                     |number         |a unique identification number for the student
|date_registration              |double         |the date of student’s registration on the module presentation|again, the weird number format that I will interpret but would need more clarification about|
|date_unregistration            |double         |date of student unregistration from the module presentation|same as above|

##### Student Info

*This file contains demographic information about the students together with their results. File contains the following columns:*

|variable                       |class          |description                    |notes                        |
|:------------------------------|:--------------|:------------------------------|:----------------------------|
|code_module                    |character      |an identification code for a module||
|code_presentation              |character      |the identification code of the presentation||
|id_student                     |number         |a unique identification number for the student||
|gender                         |character      |the student’s gender||
|region                         |character      |identifies the geographic region||
|highest_education              |character      |highest student education level on entry to the module presentation||
|imd_band                       |range (%)      |specifies the Index of Multiple Depravation band of the place where the student lived during the module-presentation||
|age_band                       |range (number) |band of the student’s age
|num_of_prev_attempts           |number         |the number times the student has attempted this module||
|studied_credits                |number         |the total number of credits for the modules the student is currently studying||
|disability                     |character      |indicates whether the student has declared a disability|(Y/N)|
|final_result                   |character      |student’s final result in the module-presentation||

##### Student VLE

*The studentVle.csv file contains information about each student’s interactions with the materials in the VLE (Virtual Learning Environment). This file contains the following columns:*

|variable                       |class          |description                    |notes                        |
|:------------------------------|:--------------|:------------------------------|:----------------------------|
|code_module                    |character      |an identification code for a module||
|code_presentation              |character      |the identification code of the presentation||
|id_student                     |number         |a unique identification number for the student||
|id_site                        |number         |an identification number for the VLE material||
|date                           |number         |the date of student’s interaction with the material||
sum_click                       |number         |the number of times a student interacts with the material on that date||

##### VLE

*The csv file contains information about the available materials in the VLE. Typically these are html pages, pdf files, etc. Students have access to these materials online and their interactions with the materials are recorded. The vle.csv file contains the following columns:*

|variable                       |class          |description                    |notes                        |
|:------------------------------|:--------------|:------------------------------|:----------------------------|
|code_module                    |character      |an identification code for a module||
|code_presentation              |character      |the identification code of the presentation||
|id_site                        |number         |an identification number for the VLE material||
|activity_type                  |character      |the role associated with the module material||
|week_from                      |number         |the week from which the material is planned to be used||
|week_to                        |number         |week until which the material is planned to be used||

#### Load in the data

In [1]:
# Load in the python modules

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
assessments_df = pd.read_csv('data/assessments.csv')
courses_df = pd.read_csv('data/courses.csv')
student_assessment_df = pd.read_csv('data/studentAssessment.csv')
student_registration_df = pd.read_csv('data/studentRegistration.csv')
student_info_df = pd.read_csv('data/studentInfo.csv')
student_vle_df = pd.read_csv('data/studentVle.csv')
vle_df = pd.read_csv('data/vle.csv')

##### Look at the first rows of every dataset

In [3]:
assessments_df.head()


code_module code_presentation  id_assessment assessment_type  date  weight
0         AAA             2013J           1752             TMA   NaN    10.0
1         AAA             2013J           1753             TMA   NaN    20.0
2         AAA             2013J           1754             TMA   NaN    20.0
3         AAA             2013J           1755             TMA   NaN    20.0
4         AAA             2013J           1756             TMA   NaN    30.0

In [4]:
courses_df.head()

code_module code_presentation  module_presentation_length
0         AAA             2113J                         268
1         AAA             2014J                         269
2         BBB             2113J                         268
3         BBB             2014J                         262
4         BBB             2013B                         240

In [5]:
student_assessment_df.head()

id_assessment  id_student  date_submitted  is_banked  score
0           1752       11391              18          0   78.0
1           1752       28400              22          0   70.0
2           1752       31604              17          0   72.0
3           1752       32885              26          0   69.0
4           1752       38053              19          0   79.0

In [6]:
student_registration_df.head()

code_module code_presentation  id_student  date_registration  \
0         AAA             2013J       11391             -159.0   
1         AAA             2013J       28400              -53.0   
2         AAA             2013J       30268              -92.0   
3         AAA             2013J       31604              -52.0   
4         AAA             2013J       32885             -176.0   

   date_unregistration  
0                  NaN  
1                  NaN  
2                 12.0  
3                  NaN  
4                  NaN

In [7]:
student_info_df.head()

code_module code_presentation  id_student gender                region  \
0         AAA             2013J       11391      M   East Anglian Region   
1         NaN               NaN       28400    NaN              Scotland   
2         AAA             2013J       30268      F  North Western Region   
3         AAA             2013J       31604      F     South East Region   
4         AAA             2013J       32885      F  West Midlands Region   

       highest_education imd_band age_band  num_of_prev_attempts  \
0       HE Qualification  90-100%     55<=                     0   
1                    NaN   20-30%      NaN                     0   
2  A Level or Equivalent   30-40%    35-55                     0   
3  A Level or Equivalent   50-60%    35-55                     0   
4     Lower Than A Level   50-60%     0-35                     0   

   studied_credits disability final_result  
0              240          N         Pass  
1               60          N         Pass  
2               60          Y    Withdrawn  
3               60          N         Pass  
4               60          N         Pass

In [8]:
student_vle_df.head()

code_module code_presentation  id_student  id_site  date  sum_click
0         AAA             2013J       28400   546652   -10          4
1         AAA             2013J       28400   546652   -10          1
2         AAA             2013J       28400   546652   -10          1
3         AAA             2013J       28400   546614   -10         11
4         AAA             2013J       28400   546714   -10          1

In [9]:
vle_df.head()

id_site code_module code_presentation activity_type  week_from  week_to
0   546943         AAA             2013J      resource        NaN      NaN
1   546712         AAA             2013J     oucontent        NaN      NaN
2   546998         AAA             2013J      resource        NaN      NaN
3   546888         AAA             2013J           url        NaN      NaN
4   547035         AAA             2013J      resource        NaN      NaN

##### Now we are going to inspect the null values

In [10]:
assessments_df.isnull().sum()

code_module            0
code_presentation      0
id_assessment          0
assessment_type        0
date                 114
weight                 0
dtype: int64

Due to the weird format of the date, at this point in time we can't make a decision about what to do with these missing values. We won't drop them for now so that we don't lose the data.

In [11]:
courses_df.isnull().sum()

code_module                   0
code_presentation             0
module_presentation_length    0
dtype: int64

In [12]:
student_assessment_df.isnull().sum()

id_assessment       0
id_student          0
date_submitted      0
is_banked           0
score             173
dtype: int64

If the student does not submit the assessment, no result is recorded.

For now we are going to change the 173 null values of score to -1.

When we are going to analyze the scores (average, mean,...), we don't want to include these rows in the analysis.


In [13]:
# Replace NaN with -1 in the score column
student_assessment_df['score'].fillna(-1, inplace=True)

In [14]:
student_registration_df.isnull().sum()

code_module                0
code_presentation          0
id_student                 0
date_registration         45
date_unregistration    22521
dtype: int64

It would be interesting to know why there are rows without registration date.

We won't remove them because they are not that many and would likely be because of an input mistake.

I will remove the registration date column but replace the unregistration date with a boolean "unregistered" so we can use this information later.

In [15]:
# Remove the date_registration column
student_registration_df.drop(columns=['date_registration'], inplace=True)

# Replace date_unregestration NaN with false and all the other values with true
student_registration_df['date_unregistration'] = student_registration_df['date_unregistration'].apply(lambda x: False if pd.isnull(x) else True)

# Rename the date_unregestration column to unregistered
student_registration_df.rename(columns={'date_unregistration': 'unregistered'}, inplace=True)

student_registration_df.head()


code_module code_presentation  id_student  unregistered
0         AAA             2013J       11391         False
1         AAA             2013J       28400         False
2         AAA             2013J       30268          True
3         AAA             2013J       31604         False
4         AAA             2013J       32885         False

In [16]:
student_info_df.isnull().sum()

code_module             3446
code_presentation       3446
id_student                 0
gender                  3446
region                     0
highest_education       3446
imd_band                1111
age_band                3446
num_of_prev_attempts       0
studied_credits            0
disability                 0
final_result               0
dtype: int64

This result is interesting because a same amount of null types appear.
Let's take a closer look at this.

In [20]:
student_info_df[student_info_df['code_module'].isnull()].sort_values("region")

code_module code_presentation  id_student gender    region  \
1             NaN               NaN       28400    NaN  Scotland   
20449         NaN               NaN      237934    NaN  Scotland   
20458         NaN               NaN      267961    NaN  Scotland   
20459         NaN               NaN      272103    NaN  Scotland   
20461         NaN               NaN      278786    NaN  Scotland   
...           ...               ...         ...    ...       ...   
11116         NaN               NaN      507031    NaN  Scotland   
11121         NaN               NaN      508637    NaN  Scotland   
11126         NaN               NaN      509710    NaN  Scotland   
11129         NaN               NaN      510266    NaN  Scotland   
32587         NaN               NaN     2620947    NaN  Scotland   

      highest_education imd_band age_band  num_of_prev_attempts  \
1                   NaN   20-30%      NaN                     0   
20449               NaN   80-90%      NaN                     0   
20458               NaN  90-100%      NaN                     1   
20459               NaN   30-40%      NaN                     0   
20461               NaN   40-50%      NaN                     1   
...                 ...      ...      ...                   ...   
11116               NaN  90-100%      NaN                     0   
11121               NaN   20-30%      NaN                     0   
11126               NaN    10-20      NaN                     0   
11129               NaN   20-30%      NaN                     0   
32587               NaN   80-90%      NaN                     0   

       studied_credits disability final_result  
1                   60          N         Pass  
20449               60          N         Fail  
20458               60          N         Fail  
20459               90          Y         Fail  
20461              135          N         Fail  
...                ...        ...          ...  
11116              120          N    Withdrawn  
11121               60          N  Distinction  
11126               30          N         Fail  
11129               60          N         Fail  
32587               30          Y  Distinction  

[3446 rows x 12 columns]

Here it is weird that every region that has missing values is Scotland, let's see for sure that this is the case.

In [21]:
scottish_students = student_info_df[student_info_df['region'] == 'Scotland']
print("--- Scottish students ---")
print("Not null values for gender:", scottish_students['gender'].notnull().sum())
print("Not null values for Highest education:", scottish_students['highest_education'].notnull().sum())
print("Not null values for Age band:", scottish_students['age_band'].notnull().sum())

print("\n--- Non-Scottish students ---")
non_scottish_students = student_info_df[student_info_df['region'] != 'Scotland']

print("Not null values for gender:", non_scottish_students['gender'].notnull().sum())
print("Null values for gender:", non_scottish_students['gender'].isnull().sum())

print("Not null values for Highest education:", non_scottish_students['highest_education'].notnull().sum())
print("Null values for highest education:", non_scottish_students['highest_education'].isnull().sum())

--- Scottish students ---
Not null values for gender: 0
Not null values for Highest education: 0
Not null values for Age band: 0

--- Non-Scottish students ---
Not null values for gender: 29147
Null values for gender: 0
Not null values for Highest education: 29147
Null values for highest education: 0


We should not drop the rows with missing values. If we drop all the rows with missing values, we lose all the Scottish students.
We would lose the final result, imd band, disablity,... of all the Scottish students.

These missing values are probably a mistake, we need to point this out to the business. They can either explain why or fix the problem.

For now, we will replace the null values with "Unknown". After the business clarifies the issue, we can revisit the problem.



In [23]:
non_scottish_students.isnull().sum()

code_module                0
code_presentation          0
id_student                 0
gender                     0
region                     0
highest_education          0
imd_band                1099
age_band                   0
num_of_prev_attempts       0
studied_credits            0
disability                 0
final_result               0
dtype: int64

There are also some students with missing values that don't live in Scotland. 
All these missing values seem to belong to imd_band. Which does not seem that weird because maybe they don't wan't to share that information.
But this is also unclear, the data could just be missing. We should also ask for clarifaction to the business.
For now, we are going to change the null values to "Missing".

In [24]:
student_info_df = student_info_df.fillna(value={"code_module": "Unknown", "code_presentation": "Unknown", "gender": "Unknown", "highest_education": "Unknown", "age_band": "Unknown", "imd_band": "Missing"})
student_info_df.isnull().sum()

code_module             0
code_presentation       0
id_student              0
gender                  0
region                  0
highest_education       0
imd_band                0
age_band                0
num_of_prev_attempts    0
studied_credits         0
disability              0
final_result            0
dtype: int64

In [17]:
student_vle_df.isnull().sum()

code_module          0
code_presentation    0
id_student           0
id_site              0
date                 0
sum_click            0
dtype: int64

In [29]:
vle_df.isnull().sum()


id_site                 0
code_module             0
code_presentation       0
activity_type           0
week_from            5243
week_to              5243
dtype: int64

5243 of the 6364 rows have missing week_from and week_to values. This probably has to do with the fact that the material is just always available or that they don't have a specific time planned for them.

We are going to leave it as null, because this seems intentional.

We should ask the business to clarify what these missing values mean.